### PR TITLE
[TUF autoupdater] Check out latest

### PR DESCRIPTION
**Pull request by RebeccaMahany**
Relates to https://github.com/kolide/launcher/issues/954.

This is the next step toward using our new TUF infrastructure: a new implementation of `findNew` that looks at our new update library.

Includes some drive-by changes:
* I've increased the download timeout from 5 to 8 minutes. I've seen just a couple download timeouts in the `kolide_tuf_autoupdater_errors` table that I'm hoping this will resolve.
* Adds a `GetChannelVersionFromTufServer` to be used by `launcher doctor` and by packaging changes for https://github.com/kolide/launcher/issues/1149

<details>
<summary>How we pick which version of launcher/osqueryd to run</summary>

This flowchart is also committed to the docs in this PR.

```mermaid
flowchart TB
    A[Library lookup] --> B{Do we have a local TUF repo?}
    B ---->|No| C[Get most recent version from update library]
    C --> D[Return path to most recent version of executable]
    D --> H[End]
    B -->|Yes| E{release.json target metadata exists?}
    E -->|No| C
    E --> |Yes| F{Target indicated by release.json\nis downloaded to update library?}
    F --> |Yes| G[Return path to selected executable in update library]
    F --> |No| C
    G --> H
```

</details>

<details>
<summary>Future work + linked PRs</summary>

Subsequent PRs will tackle the following (order is not set in stone):
1. Point to release file `<binary>/<os>/<arch>/<channel>/release.json`, once available (in progress but on hold here: https://github.com/kolide/launcher/pull/1195)
1. Trigger a reload/restart after a successful update; launcher init performs version selection using the new update library (potentially falling back to `findNew`); launcher performs library tidying after version selection
1. Other improvements/refactors (ship a more up-to-date TUF repo)
1. An "update now" functionality tied to control server
1. Eventually removing the old notary autoupdater

Previous work:
1. https://github.com/kolide/launcher/pull/1081
1. https://github.com/kolide/launcher/pull/1103
1. https://github.com/kolide/launcher/pull/1108
1. https://github.com/kolide/launcher/pull/1110
1. https://github.com/kolide/launcher/pull/1111
1. https://github.com/kolide/launcher/pull/1119
1. https://github.com/kolide/launcher/pull/1168
1. https://github.com/kolide/launcher/pull/1178
1. https://github.com/kolide/launcher/pull/1188

</details>

### CODE REVIEW

**Review by seejdev — APPROVED** [review]

**Comment by directionless** [comment]
I was slacking about this. It feels confusing and complex.

**Comment by RebeccaMahany** [comment]
> I was slacking about this. It feels confusing and complex.

@directionless I think this is simpler and more lightweight now -- let me know what you think.

File: pkg/autoupdate/tuf/library_lookup.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/kolide/launcher/pkg/autoupdate"
+)
+
+// CheckOutLatest returns the path to the latest downloaded executable for our binary, as well
+// as its version.
+func CheckOutLatest(binary autoupdatableBinary, rootDirectory string, updateDirectory string, channel string) (string, string, error) {
```
Comment (James-Pickett) [inline pkg/autoupdate/tuf/library_lookup.go:14]: I always find >2 return values confusing, consider naming the return values
```suggestion
func CheckOutLatest(binary autoupdatableBinary, rootDirectory string, updateDirectory string, channel string) (path string, version string, err error) {
```

File: pkg/autoupdate/tuf/library_lookup.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	// If we can't find the specific release version that we should be on, then just return the executable
+	// with the most recent version in the library
+	return mostRecentVersion(binary, updateDirectory)
+}
+
+// findExecutableFromRelease looks at our local TUF repository to find the release for our
+// given channel. If it's already downloaded, then we return its path and version.
+func findExecutableFromRelease(binary autoupdatableBinary, tufRepositoryLocation string, channel string, baseUpdateDirectory string) (string, string, error) {
```
Comment (James-Pickett) [inline pkg/autoupdate/tuf/library_lookup.go:31]: same consider naming return values when > 2

File: pkg/autoupdate/tuf/library_lookup.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	if autoupdate.CheckExecutable(context.TODO(), targetPath, "--version") != nil {
+		return "", "", fmt.Errorf("version %s from target %s either not yet downloaded or corrupted: %w", targetVersion, targetName, err)
+	}
+
+	return targetPath, targetVersion, nil
+}
+
+// mostRecentVersion returns the path to the most recent, valid version available in the library for the
+// given binary, along with its version.
+func mostRecentVersion(binary autoupdatableBinary, baseUpdateDirectory string) (string, string, error) {
```
Comment (James-Pickett) [inline pkg/autoupdate/tuf/library_lookup.go:59]: consider naming return values

**Review by James-Pickett — COMMENTED** [review]
I lack the understanding to review the code on a macro level. I'll try to look again later. Left some "micro" comments =)

File: pkg/autoupdate/tuf/library_lookup.go (hunk 3 above)
Comment (directionless) [reply pkg/autoupdate/tuf/library_lookup.go:59]: I agree that more than 2 return values is confusing, but I think naming them is an antipattern. (Though there's disagreement in go)

It looks pretty, but it creates them as implicit vars, which always feels like it breaks scoping to me

Comment (directionless) [reply pkg/autoupdate/tuf/library_lookup.go:59]: eg: I just accept it as imperfect. Since it feels least bad. We could talk about changing that?

File: pkg/autoupdate/tuf/library_lookup.go (hunk 1 above)
Comment (directionless) [reply pkg/autoupdate/tuf/library_lookup.go:14]: I agree that more than 2 return values is confusing, but I think naming them is an antipattern. (Though there's disagreement in go)

It looks pretty, but it creates them as implicit vars, which always feels like it breaks scoping to me

File: pkg/autoupdate/tuf/library_lookup.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// findExecutableFromRelease looks at our local TUF repository to find the release for our
+// given channel. If it's already downloaded, then we return its path and version.
+func findExecutableFromRelease(binary autoupdatableBinary, tufRepositoryLocation string, channel string, baseUpdateDirectory string) (string, string, error) {
+	// Initialize a read-only TUF metadata client to parse the data we already have downloaded about releases.
+	metadataClient, err := readOnlyTufMetadataClient(tufRepositoryLocation)
+	if err != nil {
+		return "", "", errors.New("could not initialize TUF client, cannot find release")
+	}
```
Comment (directionless) [inline pkg/autoupdate/tuf/library_lookup.go:36]: What's the behavior here if there's no local repo? (eg: first boot)

File: pkg/autoupdate/tuf/library_lookup.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	// From already-downloaded metadata, look for the release version
+	targets, err := metadataClient.Targets()
+	if err != nil {
+		return "", "", fmt.Errorf("could not get target: %w", err)
+	}
+
+	targetName, _, err := findRelease(binary, targets, channel)
+	if err != nil {
+		return "", "", fmt.Errorf("could not find release: %w", err)
+	}
```
Comment (directionless) [inline pkg/autoupdate/tuf/library_lookup.go:47]: I think this is okay, but it feels weird. These are functions that are hard to follow, and work a lot with `github.com/theupdateframework/go-tuf/data`. It feels like we should bake them into something like `metadataClient`. Not sure it's worth it though

File: pkg/autoupdate/tuf/library_lookup.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+// as its version.
+func CheckOutLatest(binary autoupdatableBinary, rootDirectory string, updateDirectory string, channel string) (string, string, error) {
+	if updateDirectory == "" {
+		updateDirectory = defaultLibraryDirectory(rootDirectory)
+	}
+
+	releasePath, releaseVersion, err := findExecutableFromRelease(binary, LocalTufDirectory(rootDirectory), channel, updateDirectory)
+	if err == nil {
+		return releasePath, releaseVersion, nil
+	}
```
Comment (directionless) [inline pkg/autoupdate/tuf/library_lookup.go:22]: Can we log the error?

**Review by directionless — COMMENTED** [review]
I think this is mostly okay. (See questions)

this feels much more readable than the prior. I suspect some of this should land in the library manager, and we should call it. Having this iterate files on disk feels wrong. (What happens if we move the library into ram?) But we can iterate there if we want to.

File: pkg/autoupdate/tuf/channel_version.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/kolide/launcher/pkg/agent"
+)
+
+// VersionForChannel returns the tagged version of the given binary for the given release channel.
+func VersionForChannel(binary, channel, tufServerUrl string) (string, error) {
```
[the remark this answers is not in the document]
Comment (RebeccaMahany) [reply pkg/autoupdate/tuf/channel_version.go:11]: What calls this:

> VersionForChannel to be used by launcher doctor and by packaging changes for https://github.com/kolide/launcher/issues/1149

Discussion on the launcher doctor PR: https://github.com/kolide/launcher/pull/1197#discussion_r1195584997

Since package build doesn't necessarily run on a machine with launcher installed, it can't use the populated TUF repo -- and either way, this feels a little safer than having separate routines updating the same TUF directory potentially at the same time.

I have a TODO to ship a more complete/up-to-date TUF repo -- hopefully that addresses the performance concern some?

File: pkg/autoupdate/tuf/library_lookup.go (hunk 4 above)
Comment (RebeccaMahany) [reply pkg/autoupdate/tuf/library_lookup.go:36]: If we can't read from the TUF repo for whatever reason, then we'd fall back to picking the most recent version in the update library. If there are no updates in the update library either, then we'd assume that the current running executable (likely the installed version, at this point) is the correct one to be running.

File: pkg/autoupdate/tuf/read_only_tuf_client.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	return r.localFilestore.Close()
+}
+
+func (r *readOnlyLocalStore) GetMeta() (map[string]json.RawMessage, error) {
+	return r.localFilestore.GetMeta()
+}
+
+func (r *readOnlyLocalStore) SetMeta(name string, meta json.RawMessage) error {
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (RebeccaMahany) [reply pkg/autoupdate/tuf/read_only_tuf_client.go:59]: I was treating these like a no-op -- I'm not sure we'd want to error out here? Erroring out would only be preventing us from executing the lookup. Same goes for https://github.com/kolide/launcher/pull/1185#discussion_r1198465017

[the remark this answers is not in the document]
Comment (directionless) [reply pkg/autoupdate/tuf/read_only_tuf_client.go:59]: Happy to defer to you!

I agree it's a no-op, but I don't know if the caller should get an error since it's an invalid action on a RO

File: pkg/autoupdate/tuf/library_lookup.go (hunk 5 above)
Comment (RebeccaMahany) [reply pkg/autoupdate/tuf/library_lookup.go:47]: I think there are a couple places in `autoupdate.go` that feel similar, where it's a lot of work with go-tuf/data. Maybe could be wrapped in a `tufReleaseClient` (or something) that performs the repeated tasks (update from remote TUF repo to download updated targets, find release target among those targets, find what target it's pointing at, return that target). I could do that here, or add it to the TODOs for a subsequent PR if we want to think about it more.

File: pkg/autoupdate/tuf/channel_version.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (directionless) [reply pkg/autoupdate/tuf/channel_version.go:11]: _ah_

So it's not used by this PR, and will be used by things that operate occasionally (package build, doctor). Given their nature, they need to talk to the server, and the additional overhead is okay. That makes sense.

I think we should name this something more clear, and godoc it appropriately. Maybe `GetVersionFromTufServer` or something?

File: pkg/autoupdate/tuf/library_lookup.go (hunk 4 above)
Comment (directionless) [reply pkg/autoupdate/tuf/library_lookup.go:36]: Perfect. I'm getting turned around because there are many functions that all call each other so tracing the flow isn't super clear to me.

File: pkg/autoupdate/tuf/channel_version.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (RebeccaMahany) [reply pkg/autoupdate/tuf/channel_version.go:11]: Yup I like that better, I will rename + add some docs!

File: pkg/autoupdate/tuf/library_lookup.go (hunk 3 above)
Comment (RebeccaMahany) [reply pkg/autoupdate/tuf/library_lookup.go]: Discussed offline, updated to return a struct

File: pkg/autoupdate/tuf/library_lookup.go (hunk 5 above)
Comment (RebeccaMahany) [reply pkg/autoupdate/tuf/library_lookup.go]: I'll leave as-is for now, I guess, and think on it

**Review by James-Pickett — APPROVED** [review]
